internal/report/source.go

```diff
@@ -1029,3 +1102,22 @@ func canonicalizeFileName(fname string) string {
 	fname = strings.TrimPrefix(fname, "./")
 	return filepath.Clean(fname)
 }
+
+// mappingList is a sorted list of mappings.
+type mappingList []*profile.Mapping
+
+func makeMappingList(src []*profile.Mapping) mappingList {
+	mappings := append([]*profile.Mapping{}, src...)
+	sort.Slice(mappings, func(i, j int) bool { return mappings[i].Start < mappings[j].Start })
+	return mappingList(mappings)
+}
+
+// find returns the mapping in m that contains addr, or nil if no such mapping exists.
+func (m mappingList) find(addr uint64) *profile.Mapping {
+	// Find first mapping that ends after addr.
+	i := sort.Search(len(m), func(i int) bool { return m[i].Limit > addr })
+	if i < len(m) && m[i].Start <= addr {
```

Comment:
I believe this makes an assumption that mappings are not overlapping in the virtual address space in the profile. This isn't necessarily true in the general case - e.g. the profile could be a merged profile of the data from multiple processes.